File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/动态规划/_02_背包问题/_02_完全背包/LeetCode_0377.java
package com.kurumi.codeThoughts.动态规划._02_背包问题._02_完全背包;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 377. 组合总和 Ⅳ
 */
public class LeetCode_0377 {
    static class Solution {
        public int combinationSum4(int[] nums, int target) {
            int[] dp = new int[target + 1];
            dp[0] = 1;
            for (int i = 0; i <= target; i++) {
                for (int num : nums) {
                    if (i >= num)
                        dp[i] += dp[i - num];
                }
            }
            return dp[target];
        }
    }
}

File: 卡码/代码随想录/1/09-动态规划/C++/dp/02-背包问题/src/02-完全背包/LeetCode_0377.cpp
//377. 组合总和 Ⅳ
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int combinationSum4(vector<int>& nums, int target) {
		vector<int> dp(target+1,0);
		dp[0]=1;
		for(int i=0;i<=target;i++){
			for(int num:nums){
				if(i>=num&&dp[i]<INT_MAX-dp[i-num]) dp[i]+=dp[i-num];
			}
		}
		return dp[target];
	}
};

int main(){
  system("pause");
  return 0;
}
